server/src-lib/Hasura/Server/Migrate.hs

```diff
@@ -123,37 +128,91 @@ migrateCatalog migrationTime = do
 
     -- migrates an existing catalog to the latest version from an existing verion
     migrateFrom :: T.Text -> m MigrationResult
-    migrateFrom previousVersion
-      | previousVersion == latestCatalogVersionString = pure MRNothingToDo
-      | [] <- neededMigrations = throw400 NotSupported $
-          "Cannot use database previously used with a newer version of graphql-engine (expected"
-            <> " a catalog version <=" <> latestCatalogVersionString <> ", but the current version"
-            <> " is " <> previousVersion <> ")."
-      | otherwise =
-          traverse_ snd neededMigrations *> postMigrate $> MRMigrated previousVersion
+    migrateFrom previousVersion =
+      case downgradeOpts of
+        Nothing
+          | previousVersion == latestCatalogVersionString ->
+              pure MRNothingToDo
+          | [] <- neededMigrations ->
+              throw400 NotSupported $
+                "Cannot use database previously used with a newer version of graphql-engine (expected"
+                  <> " a catalog version <=" <> latestCatalogVersionString <> ", but the current version"
+                  <> " is " <> previousVersion <> ")."
+          | otherwise ->
+              traverse_ (fst . snd) neededMigrations *> postMigrate $> MRMigrated previousVersion
+        Just dg
+          | previousVersion == dgoTargetVersion dg ->
+              pure MRNothingToDo
+          | otherwise -> 
+              case neededDownMigrations (dgoTargetVersion dg) of
+                Left reason -> 
+                  throw400 NotSupported $
+                    "This downgrade path (from "
+                      <> previousVersion <> " to " 
+                      <> dgoTargetVersion dg <> 
+                      ") is unfortunately not supported, because "
+                      <> reason
+                Right path -> 
+                  sequence_ path $> MRMigrated previousVersion
+      
       where
         neededMigrations = dropWhile ((/= previousVersion) . fst) migrations
+        neededDownMigrations newVersion = 
+          downgrade previousVersion newVersion (reverse migrations)
         postMigrate = updateCatalogVersion *> recreateSystemMetadata *> buildSchemaCacheStrict
 
-        migrations :: [(T.Text, m ())]
+        downgrade 
+          :: T.Text
+          -> T.Text 
+          -> [(T.Text, (m (), Maybe (m ())))]
```

Comment:
Also a minor nit, but I think this `(m (), Maybe (m ()))` type is a little non-obvious, and it shows up in a few places. Maybe pull it out into a separate `Migration` record type with field names that make it clear one of the arguments is an up migration and the other is a down migration?